### PR TITLE
NEW: Improved task runner UI.

### PR DESCRIPTION
**Pull request by mfendeksilverstripe**
# Task runner better UI

Simple UI changes which improve usability of the task runner. Lightweight CSS (no build chain or JS required).

## Problem

* it takes time to find the dev task you want to run especially if you don't know the exact task name
* links which execute dev tasks are tiny, mis-click error can happen
* current UI is cluttered with too many dev tasks and the space is used inefficiently
* rendering component of the task runner is messy (combines backend and frontend logic)
* overriding frontend display requires subclassing

### Current UI
Screen Shot 2020-06-09 at 11 14 27 AM

## Solution

* move the task into grid-style list so the screen is less cluttered
* visually separate the dev tasks
* visually separate the action buttons and make them much more visible
* split the rendering into backend (populate data) and a template (render), this allows the template to be overridden without subclassing

### Proposed UI
Screen Shot 2020-06-09 at 11 20 22 AM

## Related issues
https://github.com/silverstripe/silverstripe-framework/issues/9541

## Related Pull requests
https://github.com/symbiote/silverstripe-queuedjobs/pull/301

### CODE REVIEW

**Comment by ScopeyNZ** [comment]
Just FYI @mfendeksilverstripe, this will have to target `4` as an enhancement 🙂

**Comment by mfendeksilverstripe** [comment]
@ScopeyNZ Rebased on `4`

**Comment by emteknetnz** [comment]
Can we reduce the whitespace / padding?  Seems like it's almost been optimized for touch devices, though I cannot imagine that anyone running dev/tasks will be on an mobile device.

**Comment by mfendeksilverstripe** [comment]
@emteknetnz Padding / whitespace reduced. Please review :)

**Comment by sachajudd** [comment]
@silverstripeux to advise on UI before merge.

**Comment by ScopeyNZ** [comment]
I really struggle to understand the effort to localise this screen, or even the effort to make this responsive. I would be incredibly surprised if this page was accessed by anyone other than site developers.

**Comment by mfendeksilverstripe** [comment]
@bergice The queued job task runner UI is covered in its [own PR](https://github.com/symbiote/silverstripe-queuedjobs/pull/301)

**Comment by mfendeksilverstripe** [comment]
Not sure about about the need to have this UI responsive and localised. Would be happy to hear more detailed feedback on the consistency though.

**Comment by mfendeksilverstripe** [comment]
That's really helpful @sachajudd many thanks :)

**Comment by sachajudd** [comment]
No problem! Let me know if you need a hand, I'm happy to help out and provide feedback if needed 🙂

**Comment by mfendeksilverstripe** [comment]
@sachajudd Pushed up new changes based on UI feedback, please review. This is how the UI looks like now:

## Large screen
Screen Shot 2020-07-17 at 11 37 24 AM

## Small screen
Screen Shot 2020-07-17 at 11 37 38 AM

File: client/styles/task-runner.css
Comment (sachajudd) [inline client/styles/task-runner.css:59]: Can you add a :focus selector here too please 🙂

**Comment by sachajudd** [comment]
Hey @mfendeksilverstripe it's looking great! Awesome to see you tackle this 🙂 just noticing a few things when I pulled down your branch:

* Inconsistent spacing between some headers and descriptions
* Gaps between certain tasks

image

**Comment by mfendeksilverstripe** [comment]
@sachajudd Pushed up new changes, please review:

- added focus state for the button

I looked into the issues with inconsistent spacing between title and description. This is caused by arbitrary HTML that can be included in the task description.

Screen Shot 2020-07-17 at 1 48 12 PM

I don't think this is a layout issue. Fixing the content of specific tasks is beyond the scope of this PR.

The gaps between the tasks are caused by equal grid spacing as I'm using `display: grid`. I tried your suggestion to use `column-count` property but this can cause a single task to be split between columns. I'm open to suggestions on this one but I would be pretty happy with equal size grid cells.

**Comment by mfendeksilverstripe** [comment]
New changes:

- display now uses columns which fixes the spacing between items

@sachajudd This is now ready for review.

File: client/styles/task-runner.css
Comment (sachajudd) [inline client/styles/task-runner.css:36]: You can remove this margin-top as there a margin on the title

File: client/styles/task-runner.css
Comment (sachajudd) [inline client/styles/task-runner.css:30]: Add margin: 0 0 8px;

**Comment by mfendeksilverstripe** [comment]
@sachajudd I pushed up the changes you requested, please review.

Also, there is a handy tool you can use to suggest small changes. This is useful especially for the tiny one line changes you suggested last time ;-)

Screen Shot 2020-07-20 at 10 39 55 AM

**Comment by mfendeksilverstripe** [comment]
@sachajudd I managed to fix the spacing issue as well. It turns out there was some extra whitespace which was turned into line breaks. If you pull the latest branch you should get nice spacing for all dev tasks (between title and description).

**Comment by sachajudd** [comment]
I tried to find the commit button on Friday but couldn't seem to find it 😂 cheers, I'll take a look this afternoon

File: client/styles/task-runner.css
Comment (sachajudd) [inline client/styles/task-runner.css:56]:
```suggestion
.task__button:hover,
.task__button:focus {
```

**Comment by mfendeksilverstripe** [comment]
@Cheddam Pushed up changes as requested, please review.

**Review by Cheddam — APPROVED** [review]
Thanks for your contribution @mfendeksilverstripe!

**Pull request merged**
